chirp/drivers/ft1d.py

```diff
@@ -899,34 +902,111 @@ def _strip_ff_pads(cls, messages):
             result.append(str(msg_text).rstrip("\xFF"))
         return result
 
-    def get_memory(self, number):
-        flag = self._memobj.flag[number - 1]
-        _mem = self._memobj.memory[number - 1]
+#   slotloc called by get_memory and set_memory
+#       Called with a "memref" index to CHIRP memory (int or str)
+#   and optionally with a "extref" extended name.
+#   Find and return the corresponding memobj
+#   Returns:
+#       Corresponding radio memory object
+#       Corresponding flag structure (if any)
+#       index into overall memory (int)
+#       index into the specific memory object structure (int)
+#       an indicator of the object structure (str)
+    def slotloc(self, memref, extref=None):
+        array = None
+        num = memref
+        sname = memref
+        mstr = isinstance(memref, str)
+        specials = ALLNAMES
+        extr = False
+        if extref is not None:
+            extr = extref in specials
+        if mstr or extr:        # named special?
+            num = -1            # I'm using negative for specials
```

Comment:
Positives work mostly, but still need some help in Home specials.

Reply:
Ack. 
I'm using positive integers, but now something weird is happening... when I set a special memory, my radio structures get updated, and the display seems ok, but CHIRP complains as below. It looks to me as if _special_numbers doesn't have the index for that special register?! [1100 is the index for the first HOME register]. I don't know what I've done to bollix the _special_numbers structure. (using the "interact with driver" doesn't allow me to see what's in the wxui structures at all.
Help, please


> ERROR: <function ChirpMemEdit._memory_edited at 0x11176dee0> raised unexpected exception

> Traceback (most recent call last):
>   File "/Users/declan/chirp/chirp/wxui/common.py", line 452, in run_safe
>     return fn(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 1172, in _memory_edited
>     self.set_memory(mem)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 852, in set_memory
>     self.do_radio(set_cb, 'set_memory', chirp_common.FrozenMemory(mem))
>   File "/Users/declan/chirp/chirp/wxui/common.py", line 192, in do_radio
>     cb(job)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 846, in set_cb
>     self.do_radio(extra_cb, 'set_memory_extra', mem)
>   File "/Users/declan/chirp/chirp/wxui/common.py", line 192, in do_radio
>     cb(job)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 837, in extra_cb
>     self.refresh_memory(mem.number)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 827, in refresh_memory
>     executor(get_cb, 'get_memory', number)
>   File "/Users/declan/chirp/chirp/wxui/common.py", line 192, in do_radio
>     cb(job)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 823, in get_cb
>     executor(extra_cb, 'get_memory_extra', job.result)
>   File "/Users/declan/chirp/chirp/wxui/common.py", line 192, in do_radio
>     cb(job)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 806, in extra_cb
>     self._refresh_memory(number, job.result)
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 736, in _refresh_memory
>     row = self.mem2row(number)
>           ^^^^^^^^^^^^^^^^^^^^
>   File "/Users/declan/chirp/chirp/wxui/memedit.py", line 700, in mem2row
>     return self._special_rows[self._special_numbers[number]]
>            ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> KeyError: 1100

Reply:
You've likely got a gap or overlap in some of your special channels. The UI remembers the virtual numbers you give to your specials the first time it pulls them, so if they ever change you'll hit this.

Reply:
OK, thanks. I suppose that's possible. What field contains the "virtual numbers"??? Is it chirp.common.Memory().number ??? How does chirp.common.Memory().extd_number fit in? AFAICT, my get_memory and probably set_memory can be called with a number OR a specials name, so I must be able to handle either form and get the correct correlation.

I did a quick print test in get_memory:
With negative indices I get .number=1..900 memories, -1... -199 for specials. (0 is missing!) I get no _special_numbers problems AFAICT; it's easy to check the boundaries for the different specials. The value coming in to get_memory includes the special name, instead of a numeral or number, so the boundaries are obvious.

With positive indices, .number=1..900 memories, 901-1110 for specials. No overlaps that I see, no holes. but still the KeyError at 1100.

Reply:
`mem.number` needs to be an `int`, `mem.extd_number` is the string special name (only for special memories). `get_memory()` is called with either an `int` for regular memories or a `str` for specials. Since `set_memory()` only takes a `Memory` object, the former rules apply. By "virtual numbers" I meant the integer numbers for the specials that are greater than the end of the regular memory bounds.

If you push your code up I can possibly take a quick look, but it's hard to speculate about code I can't see...